vispy/visuals/colorbar.py

```diff
@@ -436,75 +314,166 @@ def _update(self):
+            anchors.append('bottom')
+        elif perp_direction[1] > 0:
+            anchors.append('top')
+        else:
+            anchors.append('middle')
 
-    def _prepare_transforms(self, view):
-        print("preparing transorms")
-        CompoundVisual._prepare_transforms(view)
-        view._update()
+        return anchors
 
     @staticmethod
-    def _get_orientation_error(orientation):
-        return ValueError("orientation must"
-                          " be 'horizontal' or 'vertical', "
-                          "not '%s'" % (orientation, ))
+    def _calc_positions(center, halfdim, border_width,
+                        orientation, transforms):
+        """
+        Calculate the text centeritions given the ColorBar
+        parameters.
+
+        Note
+        ----
+        This is static because in principle, this
+        function does not need access to the state of the ColorBar
+        at all. It's a computation function that computes coordinate
+        transforms
+
+        Parameters
+        ----------
+        center: tuple (x, y)
+            Center of the ColorBar
+        halfdim: tuple (halfw, halfh)
+            Half of the dimensions measured from the center
+        border_width: float
+            Width of the border of the ColorBar
+        orientation: "top" | "bottom" | "left" | "right"
+            Position of the label with respect to the ColorBar
+        transforms: TransformSystem
+            the transforms of the ColorBar
+        """
+        (x, y) = center
+        (halfw, halfh) = halfdim
+
+        visual_to_doc = transforms.get_transform('visual', 'document')
+        doc_to_visual = transforms.get_transform('document', 'visual')
+
+        # doc_widths = visual_to_doc.map(np.array([halfw, halfh, 0, 0],
+        #                                         dtype=np.float32))
+
+        doc_x = visual_to_doc.map(np.array([halfw, 0, 0, 0], dtype=np.float32))
+        doc_y = visual_to_doc.map(np.array([0, halfh, 0, 0], dtype=np.float32))
+
+        if doc_x[0] < 0:
+            doc_x *= -1
+
+        if doc_y[1] < 0:
+            doc_y *= -1
+
+        # doc_halfw = np.abs(doc_widths[0])
+        # doc_halfh = np.abs(doc_widths[1])
 
-    @staticmethod
-    def _get_anchors(orientation):
         if orientation == "top":
-            return "center", "bottom"
+            doc_perp_vector = -doc_y
         elif orientation == "bottom":
-            return "center", "top"
+            doc_perp_vector = doc_y
         elif orientation == "left":
-            return "center", "bottom"
-        else:  # orientation == "right"
-            return "center", "top"
+            doc_perp_vector = -doc_x
+        if orientation == "right":
+            doc_perp_vector = doc_x
+
+        perp_len = np.linalg.norm(doc_perp_vector)
+        doc_perp_vector /= perp_len
+        perp_len += border_width
+        perp_len *= _TEXT_PADDING_FACTOR
+        doc_perp_vector *= perp_len
```

Comment:
normalize the vector, add the border width (we're in pixel coordinates, so this is legal), scale up by the original length, add a padding factor `(_TEXT_PADDING_FACTOR = 1.05)`